presto-main/src/main/java/io/prestosql/operator/aggregation/ApproximateDoublePercentileArrayAggregations.java

```diff
@@ -96,14 +100,33 @@ public static void output(@AggregationState TDigestAndPercentileArrayState state
 
         BlockBuilder blockBuilder = out.beginBlockEntry();
 
-        for (int i = 0; i < percentiles.size(); i++) {
-            Double percentile = percentiles.get(i);
-            DOUBLE.writeDouble(blockBuilder, digest.valueAt(percentile));
+        List<Double> valuesAtPercentiles = valuesAtPercentiles(digest, percentiles);
+        for (double value : valuesAtPercentiles) {
+            DOUBLE.writeDouble(blockBuilder, value);
         }
 
         out.closeEntry();
     }
 
+    public static List<Double> valuesAtPercentiles(TDigest digest, List<Double> percentiles)
+    {
+        if (Ordering.natural().isOrdered(percentiles)) {
+            return digest.valuesAt(percentiles);
+        }
+        else {
+            // TDigest requires that percentiles list be ordered. Sort the percentiles list and rearrange the output according to original percentile order.
+            List<Double> sortedPercentiles = Ordering.natural().sortedCopy(percentiles);
+            List<Double> valuesAtPercentiles = digest.valuesAt(sortedPercentiles);
+            Map<Double, Double> percentilesToValues = new HashMap<>();
+            for (int i = 0; i < sortedPercentiles.size(); i++) {
+                percentilesToValues.put(sortedPercentiles.get(i), valuesAtPercentiles.get(i));
+            }
+            return percentiles.stream()
+                    .map(percentilesToValues::get)
+                    .collect(toImmutableList());
```

Comment:
This may be more efficient with fastutil's indirect sorting (you should give it a try). Something like:

```java
int[] indexes = new int[percentiles.size()];
double[] sortedPercentiles = new double[percentiles.size()];
for (int i = 0; i < indexes.length; i++) {
    indexes[i] = i;
    sortedPercentiles[i] = percentiles.get(i);
}

it.unimi.dsi.fastutil.Arrays.quickSort(0, percentiles.size(), (a, b) -> Doubles.compare(sortedPercentiles[a], sortedPercentiles[b]), (a, b) -> {
    double tempPercentile = sortedPercentiles[a];
    sortedPercentiles[a] = sortedPercentiles[b];
    sortedPercentiles[b] = tempPercentile;

    int tempIndex = indexes[a];
    indexes[a] = indexes[b];
    indexes[b] = tempIndex;
});

List<Double> valuesAtPercentiles = digest.valuesAt(Doubles.asList(sortedPercentiles));
List<Double> result = new ArrayList<>(valuesAtPercentiles.size());
for (int i = 0; i < valuesAtPercentiles.size(); i++) {
    result.add(valuesAtPercentiles.get(indexes[i]));
}

return result;
```